### PR TITLE
core.thread refactor: osthread module

### PR DESCRIPTION
**Pull request by baziotis**
This is the start of an effort to refactor `core.thread`. With Nicholas Wilson's help, this takes the form of a separate folder that will split `thread` to modules. Currently, the plan is to split it to:
- `osthread` : Low-level, OS-dependent code for threads.
- `fiber` : The interface for fibers.
- `context` : The stack context needed for threads and fibers.

`osthread` seems to be the easiest to start from so that the review is easier.

### CODE REVIEW

**Comment by thewilsonator** [comment]
You should start by moving `core/thread.d` into `core/thread/package.d` and then move things out of there into other files, making `private` symbols `package(core.thread)` where needed. 

It will be much easier to keep track of things that way.

**Comment by baziotis** [comment]
Ok, I just didn't get the last phrase:
> making private symbols package(core.thread) where needed.

**Comment by thewilsonator** [comment]
some symbols in core.thread are private. 

When you move symbol around that require access to those symbols, as they are now in different modules they will give an error if used. 

making those symbols `package(core.thread)` will resolve that problem and still have the symbols not be visible outside of `core.thread`

**Comment by baziotis** [comment]
Ok, thanks. I didn't know that there's the `package` specifier. I'll learn about it and update the PR.

**Comment by ZombineDev** [comment]
@baziotis thanks for working on this! I will try to review your later today or tomorrow.

**Comment by baziotis** [comment]
@ZombineDev Don't mention it. It might seem like a big PR but it's actually not. I just moved the OS-dependent stuff from `core/thread.d` to `/thread/osthread.d`. I'll continue with Nicholas' proposal.

**Comment by Geod24** [comment]
> I'll continue with Nicholas' proposal.

@thewilsonator 's proposal was to *start* with something else. And IMO it makes a lot of sense.
This PR is currently only adding code, nothing is removed yet, so I assume this is in essence unused code ? Which also means untested. It's harder to review when the changes are actually disconnected, like adding in a PR and removing in another (unless there's a good reason to do so, e.g. when dealing with cross-repository dependencies).

**Comment by baziotis** [comment]
@Geod24 To be honest, I'm not sure what is your point. This PR is a draft because it is not supposed to be merged as it is. It basically asks "have I moved the correct amount of things (i.e. the OS stuff) from `thread.d` to `osthread`". Practically, it adds no code.

**Comment by thewilsonator** [comment]
> have I moved the correct amount of things

In this case, no, because there should be a (roughly) equal amount of code removed.

File: src/core/thread/package.d [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+ * License: Distributed under the
+ *      $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost Software License 1.0).
+ *    (See accompanying file LICENSE)
+ * Authors:   Sean Kelly, Walter Bright, Alex Rønne Petersen, Martin Nowak
+ * Source:    $(DRUNTIMESRC core/thread/package.d)
+ */
+
+module core.thread;
+
+public import core.time;  // Should be removed
```
Comment (thewilsonator) [inline src/core/thread/package.d]: can these be `package(core.thread) import`s?

Comment (baziotis) [reply src/core/thread/package.d]: Let me post the PM:

The initial `thread.d` had `public import core.time`. So, I removed `public` but then `spinlock` (iirc) used something of `core.time` which was coming from `thread`.
It was kind of a mess so I removed the `public` from `core.thread` and imported on `spinlock`.
As you can see, I get an `undefined reference` error which I don't understand. Thinking that it might have been a problem of with removing the `public import` from `core.thread`, I put it again on `package.d`.

Tl;dr, the `public import core.time` should not be needed. The rest yes, they can.

**Comment by baziotis** [comment]
Can I have some help on why the tests are failing? I can't seem to find a way to fix them.

**Comment by aG0aep6G** [comment]
> Can I have some help on why the tests are failing? I can't seem to find a way to fix them.

I think you forgot to add thread/package.d to the make files.

**Comment by baziotis** [comment]
> I think you forgot to add thread/package.d to the make files.
I thought it wasn't needed, thanks for the response. I added in all but DOCS. Unfortunately, it doesn't seem to fix the error.

**Comment by aG0aep6G** [comment]
> Unfortunately, it doesn't seem to fix the error.

The error is different now. [You've got a merge conflict in mak/DOCS.](https://auto-tester.puremagic.com/show-run.ghtml?projectid=1&runid=3712300&isPull=true) You need to rebase.

**Comment by baziotis** [comment]
> The error is different now. You've got a merge conflict in mak/DOCS. You need to rebase.

Yes, I had this error in the previous commit as well. But, besides this, there's also the previous error for example in circleci.

**Comment by thewilsonator** [comment]
[This test](https://github.com/dlang/dmd/blob/master/test/runnable/test15779.d) seems to fail on windows consistently.

File: src/core/thread/osthread.d [hunk 2]
```diff
@@ -1579,8 +1648,27 @@ private:
         sm_this = t;
     }
 
+package(core.thread):
```
Comment (thewilsonator) [inline src/core/thread/osthread.d]: One of the reasons for wanting to do this refactoring in the first place was to have this be pubic so that Weka can use it in their own fiber code. Please remove this line.

Comment (thewilsonator) [reply src/core/thread/osthread.d]: and move it outside of `class Thread`

**Comment by baziotis** [comment]
> [This test](https://github.com/dlang/dmd/blob/master/test/runnable/test15779.d) seems to fail on windows consistently.

Thanks, I currently can't build either dmd or druntime on Windows. With either Visual Studio or the makefiles. Building those from source can have a lot of complications.

**Comment by baziotis** [comment]
I still don't understand the error much. It seems it has something to do with [the constructor of Fiber](https://github.com/dlang/druntime/blob/e36084b951eb94b382bd0a6b253372e51f5c4c4d/src/core/thread/fiber.d#L546) or [call()](https://github.com/dlang/druntime/blob/e36084b951eb94b382bd0a6b253372e51f5c4c4d/src/core/thread/fiber.d#L637).

The second case AFAIK is harder since it goes quite deep in the calls. However for the first, could you please verify that the module construction works across the package ? Meaning [this](https://github.com/dlang/druntime/blob/e36084b951eb94b382bd0a6b253372e51f5c4c4d/src/core/thread/osthread.d#L76).

**Comment by baziotis** [comment]
Finally that works. :sweat_smile: 
@thewilsonator Regarding `Context`. I'm not sure whether moving `Context` outside is better done in this PR. It might not be enough to just move the struct and also I liked the version in your diff where there was a separate file for all the context stuff.

**Comment by thewilsonator** [comment]
I take it this is no linger a draft PR?

**Comment by baziotis** [comment]
> I take it this is no longer a draft PR?

Yes, let me make it an actual PR.

**Comment by baziotis** [comment]
Oh crap, I thought `Ready for review` makes it an actual PR, not requesting actual reviews.

**Comment by lesderid** [comment]
> Edit: Should this somehow be undone or is it supposed to be that way?

Supposed to be that way.

**Comment by thewilsonator** [comment]
There are a couple of things I'd like to see changed, but theres no point keeping thins waiting any further.

**Comment by baziotis** [comment]
> There are a couple of things I'd like to see changed, but theres no point keeping thins waiting any further.

You can tell me. :)
As this is a PR in a series, trying to break the changes in reasonable PRs.
To the best of my knowledge, I have done the things that were discussed in this one.

**Comment by thewilsonator** [comment]
https://github.com/dlang/druntime/pull/2689#discussion_r312720591
Fair enough.
Also fair enough.

For the next one split out the context stuff into `thread/context.d` like in the diff I gave you. (it has some typos and naming inconsistencies but I'm sure you'll discover that when you try to compile it ;) )

**Comment by baziotis** [comment]
> [#2689 (comment)](https://github.com/dlang/druntime/pull/2689#discussion_r312720591)
> Fair enough.
> Also fair enough.
> 

Yes as said here: https://github.com/dlang/druntime/pull/2689#issuecomment-520237162
I think it makes this easier to do in a separate PR.

> For the next one split out the context stuff into `thread/context.d` like in the diff I gave you. (it has some typos and naming inconsistencies but I'm sure you'll discover that when you try to compile it ;) )

Haha yes I don't compile it. As you can see from this PR there are a lot of subtle things
in these changes so I take your idea and try to figure out how to make the implementation work.

**Comment by thewilsonator** [comment]
Indeed, I look forward to it.

**Comment by rainers** [comment]
As one of the tests failing in https://dev.azure.com/dlanguage/dmd/_build/results?buildId=4590 and other PRs is runnable/test15779.d which checks exceptions in fibers, I suspect it has been introduced by this PR. Any idea what might be causing this?

Apart from the auto tester, druntime just runs a partial win64 test on Appveyor. @wilzbach Can we easily run the azure tests from dmd for druntime, too?

**Comment by thewilsonator** [comment]
No, but I note that there are a [couple](https://github.com/dlang/druntime/blob/master/src/core/thread/fiber.d#L1721) of [unittests](https://github.com/dlang/druntime/blob/master/src/core/thread/fiber.d#L1761) related to exceptions and fibers that are explicitly disabled for Win32. Maybe a subtle change has caused it to hit that bug?

**Comment by rainers** [comment]
Indeed, the disabled unittests look very similar. I have failed to reproduce the issues reported by the azure CI locally, will try the unittests tonight.

**Comment by thewilsonator** [comment]
Of note, that test enables stack stomping.

**Comment by baziotis** [comment]
Follow-up: https://github.com/dlang/druntime/pull/2821
